specs/Deployment/DeploymentAPI.md

```diff
@@ -111,7 +111,55 @@ again.
 ```C#
     if (DeploymentManager.GetStatus().Status != DeploymentStatus.Ok)
     {
-        // Initialize does a status check, and if the status is not Ok it will attempt to get
+        // Initialize does a status check, and if the status is not OK it will attempt to get
+        // the WindowsAppRuntime into a good state by deploying packages. Unlike a simple
+        // status check, Initialize can sometimes take several seconds to deploy the packages.
+        // These should be run on a separate thread so as not to hang your app while the
+        // packages deploy.
+        var initializeTask = Task.Run(() => DeploymentManager.Initialize());
+        initializeTask.Wait();
+
+        // Check the result.
+        if (initializeTask.Result.Status != DeploymentStatus.Ok)
+        {
+            // The WindowsAppRuntime is in a bad state which Initialize() did not fix.
+            // Do error reporting or gather information for submitting a bug.
+            // Gracefully exit the program or carry on without using the WindowsAppRuntime.
+        }
+    }
+```
+
+### DeploymentInitializeOptions
+
+#### ForceTargetApplicationShutDown
+
+When this API may install newer version of one or more of the packages (i.e. an update) and if there
+are any currently running process(es) associated with that package(s) (in other words, if any of the
+packages to be updated are currently in use), then this API will fail installing that package. But
+this update to the packages can be forced by using DeploymentInitializeOptions object and setting
+ForceTargetApplicationShutDown option before passing it to this API. This option when set will
+shutdown the application(s) associated with the package, update the package and restart the
+application(s).
```

Comment:
Wait, it restarts them too? The property name said it was just going to shut them down.

Should this be ForceRestartApps?

Or should we look at it from the point of view of the update. "UpdateEvenIfInUse" / "ForceUpdateNow"?

Reply:
> Or should we look at it from the point of view of the update. "UpdateEvenIfInUse" / "ForceUpdateNow"?

That's the better question. We can add framework packages to a system and don't interrupt processes using them, but main packages cannot be serviced while in use. Running processes using a main package block deployment, unless a 'force' option was specified in which case we TerminateProcess, do the work, then restart any _packaged apps_ we terminated.

Not all processes are packaged apps (e.g. Packaged COM OOP Servers don't get restarted, and of course ditto unpackaged apps). _Today_ there's no impact but YMMV tomorrow (there's 1 case today the OS doesn't automagically restart, but the API will do the work to handle it).

Assume we're writing the PackageManager API for the 1st time. What would you call an option to "force the deployment operation regardless what's running, terminating processes and/or changing machine state as needed. The deployment request is top priority, make it happen, impacts to running processes is of lower priority" ?

That's generally viewed as a "force it" option. The "and after restore anything you did before" isn't 100%. Packaged apps can be restarted but not other types of processes. ForceTheDeploymentOperationAndIfAnyProcessWasTerminatedRestoreAsBestEffort? :P